pkg/resource/default.go

```diff
@@ -0,0 +1,121 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resource
+
+import (
+	"context"
+	"strings"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	corev1alpha1 "github.com/crossplaneio/crossplane/pkg/apis/core/v1alpha1"
+	"github.com/crossplaneio/crossplane/pkg/logging"
+)
+
+const (
+	controllerNameDefault = "defaultclass.crossplane.io"
+)
+
+var logDefault = logging.Logger.WithName("controller").WithValues("controller", controllerNameDefault)
+
+// Error strings
+const (
+	errNoDefault        = "unable to locate a default resource class for claim kind"
+	errMultipleDefaults = "multiple default classes defined for claim kind"
+	errInvalidLabel     = "unable to set label for claim kind"
+)
+
+// DefaultReconciler reconciles resource claims to the
+// default resource class for their given kind. Predicates
+// ensure that only claims with no resource class reference
+// are reconciled.
+type DefaultReconciler struct {
+	client   client.Client
+	newClaim func() Claim
+}
+
+// NewDefaultReconciler creates a new DefaultReconciler for the claim kind
+func NewDefaultReconciler(m manager.Manager, of ClaimKind) *DefaultReconciler {
+	nc := func() Claim { return MustCreateObject(schema.GroupVersionKind(of), m.GetScheme()).(Claim) }
+
+	_ = nc()
+
+	r := &DefaultReconciler{
+		client:   m.GetClient(),
+		newClaim: nc,
+	}
+
+	return r
+}
+
+// Reconcile reconciles a claim to the default class reference for its kind
+func (r *DefaultReconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	logDefault.V(logging.Debug).Info("Reconciling", "request", req)
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	claim := r.newClaim()
+	if err := r.client.Get(ctx, req.NamespacedName, claim); err != nil {
+		// There's no need to requeue if we no longer exist. Otherwise we'll be
+		// requeued implicitly because we return an error.
+		return reconcile.Result{}, errors.Wrap(IgnoreNotFound(err), errGetClaim)
+	}
+
+	// Get claim group and kind information
+	claimKind := claim.GetObjectKind()
+	claimGroupVersionKind := claimKind.GroupVersionKind()
+	claimGroupKind := claimGroupVersionKind.GroupKind()
+	claimGroupKindStr := strings.ToLower(claimGroupKind.String())
+
+	// Create label selector in crossplane-system namespace
+	options := client.InNamespace("crossplane-system")
```

Comment:
We should not hardcode this namespace. It's only convention today that classes are created in `crossplane-system` - it's possible to create them in any namespace. Can we use a label selector against classes from all namespaces?

Reply:
Yes, I believe so. `client.InNamespace` returns `*client.ListOptions`, but we can instantiate an empty `ListOptions` struct, `options := &client.ListOptions{}`, then set the Label Selector on it.